## promise的三种状态

-   promise是一种异步的操作，它有三种状态,Pending（进行中）、Resolved（已完成，又称Fulfilled）和Rejected（已失败)，实际变化只有下面两种状态,一旦改变就不可以修改。
-   Pending ->  Resolved
-   Pending -> Rejected

![](https://p3-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/abe9b980bd4947ffbc0379feb7f50696~tplv-k3u1fbpfcp-zoom-1.image)

## 例子

```js
new Promise(function(resolve, reject){
    const a = 1;
    if(a === 1){
        resolve(a)
    }else{
        reject('error)
    }
}).then(function(value){
    console.log(value)
}).catch(function(error){
    console.log(error)
})
```

## finally

-   finally 方法用于指定不管 Promise 对象最后状态如何，都会执行的操作。finally 方法的回调函数不接受任何参数。

```js
let profiy = new Promise((resolve,reject) => {
    if (true) {
        resolve('哈哈哈');
    }
}).then(res => {
    console.log('res:'+ res);
}).catch(error => {
    console.log('errpr:' + error);
}).finally(()=> {
    console.log('呵呵呵');
})
// 哈哈哈  呵呵呵
```

## promise.all()

-   Promsie.all() 方法用于将多个 Promise 实例包装成一个新的 Promise 实例。

```js
let p = Promise.all([p1,p2,p3]);
```

-   p的状态是由p1,p2,p3共同决定,有两种情况

<!---->

-   -   当p1,p2,p3的状态都是成功时，p的状态才是成功的,此时p1,p2,p3的返回值会组成一个数组,返回给p
    -   当p1,p2,p3的状态有一个失败时，p的状态也是失败，如果p1,p2,p3有自己的catch就会执行自己的，没有则执行p的catch方法。

```js
//都成功情况
let p1 = new Promise((resolve,reject)=>{
    resolve('p1');
})

let p2 = new Promise((resolve,reject)=>{
    resolve('p2');
})

Promise.all([p1,p2]).then(res=>{
    console.log(res);
}).catch(error=>{
    console.log('error:' + error);
})

// ['p1', 'p2']
```

```js
//失败情况
let p1 = new Promise((resolve,reject)=>{
    resolve('p1');
})

let p2 = new Promise((resolve,reject)=>{
    // resolve('p2');
    throw new Error('报错了');
}).catch(error=>{
    console.log('p2 error:' + error);
})

Promise.all([p1,p2]).then(res=>{
    console.log(res);
}).catch(error=>{
    console.log(error);
})
//Error: 报错了	//Promise.all()有自己的catch,会执行Promise.all()的，如果没有则执行自己的catch
```

## promise.race()

-   Promise.race() 方法也是将多个 Promise 实例，包装成一个新的 Promise 实例.
-   如果p1,p2,p3中有一个实例先改变，那么最先改变的状态的实例返回值就会传递给p,谁先执行完就会返回谁的值。

```js
const p = Promise.race([p1, p2, p3]);

//实例
let p1 = new Promise((resolve,reject)=>{
    resolve('p1');
})

let p2 = new Promise((resolve,reject)=>{
    resolve('p2');
}).catch(error=>{
    console.log('p2 error:' + error);
})

Promise.race([p1,p2]).then(res=>{
    console.log(res);
})

//p1		p1先执行，将数据返回到方法的回调函数中去。
```

## promise.allSettled()

Promise.allSettled()方法接收一个Promise对象的数组作为参数，并返回一个新的Promise对象。这个新的Promise对象会在所有Promise对象都完成后被resolved，不管这些Promise对象是成功还是失败，它们的结果都会被收集到一个数组中返回。

如果所有的Promise对象都成功了，这个数组的每个元素都是一个对象，它有两个属性：status和value。status的值为"fulfilled"，value的值为对应Promise对象的返回值。

如果有任何一个Promise对象失败了，这个数组的每个元素都是一个对象，它有两个属性：status和reason。status的值为"rejected"，reason的值为对应Promise对象失败的原因。

Promise.allSettled()方法的返回值是一个Promise对象，它的状态只能是resolved。如果所有的Promise对象都成功了，这个Promise对象的返回值是一个数组，数组里的每个元素都是一个对象，它的结构如上所述。如果有任何一个Promise对象失败了，这个Promise对象的返回值也是一个数组，数组里的每个元素也是一个对象，它的结构如上所述。

```js
const promise1 = Promise.resolve(3);
const promise2 = new Promise((resolve, reject) => setTimeout(reject, 100, 'foo'));
const promises = [promise1, promise2];

Promise.allSettled(promises).
  then((results) => results.forEach((result) => console.log(result)));

{ status: 'fulfilled', value: 3 }
{ status: 'rejected', reason: 'foo' }
```

## promise.any()

promise.any()方法接收一个Promise对象的数组作为参数，并返回一个新的Promise对象。这个新的Promise对象会在这些Promise对象中有任何一个变为fulfilled状态时被resolved，并返回该Promise对象的值。

如果所有的Promise对象都失败了，这个新的Promise对象会被rejected，并返回一个AggregateError对象，它包含所有Promise对象的失败原因。

Promise.any()方法与Promise.race()方法类似，但是它只要求有一个Promise对象变为fulfilled状态即可，而不是要求所有Promise对象都完成。与Promise.race()方法不同的是，如果其中一个Promise对象失败了，Promise.race()方法会直接返回该Promise对象的失败原因，而Promise.any()方法会等待所有Promise对象都失败后，将所有失败原因收集起来返回。

```js
const pErr = new Promise((resolve, reject) => {
  reject("总是失败");
});

const pSlow = new Promise((resolve, reject) => {
  setTimeout(resolve, 500, "最终完成");
});

const pFast = new Promise((resolve, reject) => {
  setTimeout(resolve, 100, "很快完成");
});

Promise.any([pErr, pSlow, pFast]).then((value) => {
  console.log(value);
})

// 很快完成
```

## promise封装ajax

```js
const getJSON = function(url) {
      const promise = new Promise(function(resolve, reject){
        const handler = function() {
          if (this.readyState !== 4) {
            return;
          }
          if (this.status === 200) {
            resolve(this.response);
          } else {
            reject(new Error(this.statusText));
          }
        };
        const client = new XMLHttpRequest();
        client.open("GET", url);
        client.onreadystatechange = handler;
        client.responseType = "json";
        client.setRequestHeader("Accept", "application/json");
        client.send();
      });
      return promise;
    };
```